### PR TITLE
24 bit FLAC export. Clip negative side of wave

### LINKED ISSUE #5128: FLAC export not identical to WAV
When exporting to FLAC with a recent master build, I noticed that the output had noticeable (sometimes severe) crackling not present in previous exports. Here's a comparison of the output waveform, with WAV on top and FLAC on the bottom:

image

Examining the top left of both waveforms, the FLAC export appears to have a downwards DC offset. In addition, when this offset wave reaches -1.0, it wraps around and "sticks" at 1.0.

Export options: 44.1kHz, 24 Bit, Sinc best, No oversampling

Also, a minor filename bug:

- When choosing output location, leave extension on the default "WAV"
- Select FLAC as output format
- File extension of exported file is `.wav.flac`

### PR DESCRIPTION
**Pull request by zonkmachine**
Fixes #5128

### CODE REVIEW

File: src/core/audio/AudioFileFlac.cpp [hunk 1]
```diff
@@ -94,7 +96,10 @@ void AudioFileFlac::writeBuffer(surroundSampleFrame const* _ab, fpp_t const fram
 		{
 			for(ch_cnt_t channel=0; channel<channels(); ++channel)
 			{
-				buf[frame*channels() + channel] = _ab[frame][channel] * master_gain;
+				// Clip the negative side to -0.999999 in order to prevent it from changing sign
+				// Upstream issue: https://github.com/erikd/libsndfile/issues/309
+				// When this commit is reverted libsndfile-1.0.29 must be made a recuirement for FLAC
+				buf[frame*channels() + channel] = qMax(-0.999999f, _ab[frame][channel] * master_gain);
```
Comment (PhysSong) [inline src/core/audio/AudioFileFlac.cpp:102]: `std::nextafterf(-1.0f, 0.0f)` from `<cmath>` is a portable alternative of the constant.

Comment (zonkmachine) [reply src/core/audio/AudioFileFlac.cpp]: Cool! I see that I used nextafter() instead of nextafterf(). They seem to be the same though.

Comment (zonkmachine) [reply src/core/audio/AudioFileFlac.cpp]: Same-ish. Changing to nextafterf().

**Comment by zonkmachine** [comment]
@PhysSong I'm happy with this now. It would probably be a good idea if someone else tested it. Just render a short note, that is overdriven and clipping in lmms, in 24 bit flac and make sure it's not wonky.

File: src/core/audio/AudioFileFlac.cpp [hunk 2]
```diff
@@ -94,7 +98,10 @@ void AudioFileFlac::writeBuffer(surroundSampleFrame const* _ab, fpp_t const fram
 		{
 			for(ch_cnt_t channel=0; channel<channels(); ++channel)
 			{
-				buf[frame*channels() + channel] = _ab[frame][channel] * master_gain;
+				// Clip the negative side to -0.999999 in order to prevent it from changing sign
```
Comment (zonkmachine) [inline src/core/audio/AudioFileFlac.cpp:101]: Should this now be `// Clip the negative side to just above -1.0 ...` ?

Comment (Spekular) [reply src/core/audio/AudioFileFlac.cpp:101]: I think so.

File: src/core/audio/AudioFileFlac.cpp [hunk 3]
```diff
@@ -94,7 +98,10 @@ void AudioFileFlac::writeBuffer(surroundSampleFrame const* _ab, fpp_t const fram
...
+				// Upstream issue: https://github.com/erikd/libsndfile/issues/309
+				// When this commit is reverted libsndfile-1.0.29 must be made a recuirement for FLAC
```
Comment (Spekular) [inline src/core/audio/AudioFileFlac.cpp:103]:
```suggestion
				// When this commit is reverted libsndfile-1.0.29 must be made a requirement for FLAC
```

File: src/core/audio/AudioFileFlac.cpp (hunk 2 above)
Comment (zonkmachine) [reply src/core/audio/AudioFileFlac.cpp]: Fixed!

**Comment by zonkmachine** [comment]
@PhysSong and @Spekular Thanks for the review and testing!

I believe this is ready to merge. We found a noise issue, a small difference between exporting wav and flac (both at 24 bit/44.1kHz) that seemed to be not related to this PR. It looks like a one bit rounding off error that is distributed between the two waves except where clipping occurs. Where there is clipping there is no difference between the sounds. It should be documented though.

**Comment by zonkmachine** [comment]
> It looks like a one bit rounding off error that is distributed between the two waves except where clipping occurs.

Note to self. Audacity performs dither by default on some sample conversions. Must turn it off and redo the tests. In this case specifically the dither could have occurred on loading samples into Audacity.

**Comment by zonkmachine** [comment]
I can't replicate the noise with 16 bit export but the noise is there on master on 24 bit. Just export some file that isn't clipping and it's there already. Turning off dither on audacity didn't do anything and I also tried only bitwidth in the sound chain that wouldn't dither and that didn't do it either. Probably not dither then.

**Comment by zonkmachine** [comment]
I think we're done here. Merge?

**Comment by Spekular** [comment]
Agreed!

**Pull request merged**